ext/native/thin3d/thin3d_gl.cpp

```diff
@@ -1233,595 +967,181 @@ void OpenGLContext::UpdateDynamicUniformBuffer(const void *ub, size_t size) {
-			switch (desc.attributes[i].format) {
-			case DataFormat::R32G32_FLOAT:
-				glVertexAttribPointer(desc.attributes[i].location, 2, GL_FLOAT, GL_FALSE, stride, (void *)(b + (intptr_t)desc.attributes[i].offset));
-				break;
-			case DataFormat::R32G32B32_FLOAT:
-				glVertexAttribPointer(desc.attributes[i].location, 3, GL_FLOAT, GL_FALSE, stride, (void *)(b + (intptr_t)desc.attributes[i].offset));
-				break;
-			case DataFormat::R32G32B32A32_FLOAT:
-				glVertexAttribPointer(desc.attributes[i].location, 4, GL_FLOAT, GL_FALSE, stride, (void *)(b + (intptr_t)desc.attributes[i].offset));
-				break;
-			case DataFormat::R8G8B8A8_UNORM:
-				glVertexAttribPointer(desc.attributes[i].location, 4, GL_UNSIGNED_BYTE, GL_TRUE, stride, (void *)(b + (intptr_t)desc.attributes[i].offset));
-				break;
-			case DataFormat::UNDEFINED:
-			default:
-				ELOG("Thin3DGLVertexFormat: Invalid or unknown component type applied.");
-				break;
-			}
-		}
-		if (id_ != 0) {
-			lastBase_ = b;
-		}
-	}
-}
-
-void OpenGLInputLayout::Unapply() {
-	if (id_ == 0) {
-		for (int i = 0; i < (int)SEM_MAX; i++) {
-			if (semanticsMask_ & (1 << i)) {
-				glDisableVertexAttribArray(i);
-			}
-		}
-	} else {
-		glBindVertexArray(0);
-	}
+OpenGLInputLayout::~OpenGLInputLayout() {
 }
 
-// On PC, we always use GL_DEPTH24_STENCIL8. 
-// On Android, we try to use what's available.
+void OpenGLInputLayout::Compile(const InputLayoutDesc &desc) {
+	int semMask = 0;
 
-#ifndef USING_GLES2
-OpenGLFramebuffer *OpenGLContext::fbo_ext_create(const FramebufferDesc &desc) {
-	OpenGLFramebuffer *fbo = new OpenGLFramebuffer();
-	fbo->width = desc.width;
-	fbo->height = desc.height;
-	fbo->colorDepth = desc.colorDepth;
-
-	// Color texture is same everywhere
-	glGenFramebuffersEXT(1, &fbo->handle);
-	glGenTextures(1, &fbo->color_texture);
-
-	// Create the surfaces.
-	glBindTexture(GL_TEXTURE_2D, fbo->color_texture);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
-	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
-
-	// TODO: We could opt to only create 16-bit render targets on slow devices. For later.
-	switch (fbo->colorDepth) {
-	case FBO_8888:
-		glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, fbo->width, fbo->height, 0, GL_RGBA, GL_UNSIGNED_BYTE, NULL);
-		break;
-	case FBO_4444:
-		glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, fbo->width, fbo->height, 0, GL_RGBA, GL_UNSIGNED_SHORT_4_4_4_4, NULL);
-		break;
-	case FBO_5551:
-		glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, fbo->width, fbo->height, 0, GL_RGBA, GL_UNSIGNED_SHORT_5_5_5_1, NULL);
-		break;
-	case FBO_565:
-		glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, fbo->width, fbo->height, 0, GL_RGB, GL_UNSIGNED_SHORT_5_6_5, NULL);
-		break;
-	}
+	// This is only accurate if there's only one stream. But whatever, for now.
```

Comment:
Hmm, should we at least put a TODO here?

-[Unknown]

Reply:
Yup